brainiak/reprsimil/brsa.py

```diff
@@ -867,38 +1324,258 @@ def _fit_RSA_UV(self, X, Y, X0,
+            est_std_log_SNR, est_intensity_kernel_r, X0
+
+    def _transform(self, Y, scan_onsets, beta, beta0,
+                   rho_e, sigma_e, rho_X, sigma2_X, rho_X0, sigma2_X0):
+        """ Given the data Y and the response amplitudes beta and beta0
+            estimated in the fit step, estimate the corresponding X and X0.
+            It is done by a forward-backward algorithm.
+            We assume X and X0 both are vector autoregressive (VAR)
+            processes, to capture temporal smoothness. Their VAR
+            parameters are estimated from training data at the fit stage.
+        """
+        logger.info('Transforming new data.')
+        # Constructing the transition matrix and the variance of
+        # refreshing noise as prior for the latent variable X and X0
+        # in new data.
+
+        n_C = beta.shape[0]
+        n_T = Y.shape[0]
+        weight = np.concatenate((beta, beta0), axis=0)
+        T_X = np.diag(np.concatenate((rho_X, rho_X0)))
+        Var_X = np.concatenate((sigma2_X / (1 - rho_X**2),
+                                sigma2_X0 / (1 - rho_X0**2)))
+        Var_dX = np.concatenate((sigma2_X, sigma2_X0))
+        sigma2_e = sigma_e ** 2
+        scan_onsets = np.setdiff1d(scan_onsets, n_T)
+        n_scan = scan_onsets.size
+        X = [None] * scan_onsets.size
+        X0 = [None] * scan_onsets.size
+        total_log_p = 0
+        for scan, onset in enumerate(scan_onsets):
+            # Forward step
+            if scan == n_scan - 1:
+                offset = n_T
+            else:
+                offset = scan_onsets[scan + 1]
+            mu, mu_Gamma_inv, Gamma_inv, log_p_data, Lambda_0, \
+                Lambda_1, H, deltaY, deltaY_sigma2inv_rho_weightT = \
+                self._forward_step(Y[onset:offset, :],
+                                   T_X, Var_X, Var_dX, rho_e, sigma2_e,
+                                   weight)
+            total_log_p += log_p_data
+            # Backward step
+            mu_hat, mu_Gamma_inv_hat, Gamma_inv_hat \
+                = self._backward_step(
+                    deltaY, deltaY_sigma2inv_rho_weightT, sigma2_e,
+                    weight, mu, mu_Gamma_inv, Gamma_inv,
+                    Lambda_0, Lambda_1, H)
+            X[scan] = np.concatenate(
+                [mu_t[None, :n_C] for mu_t in mu_hat])
+            X0[scan] = np.concatenate(
+                [mu_t[None, n_C:] for mu_t in mu_hat])
+        X = np.concatenate(X)
+        X0 = np.concatenate(X0)
+        return X, X0, total_log_p
+
+    def _score(self, Y, design, beta, scan_onsets, beta0, rho_e, sigma_e,
+               rho_X0, sigma2_X0):
+        """ Given the data Y, and the spatial pattern beta0
+            of nuisance time series, return the cross-validated score
+            of the data Y given all parameters of the subject estimated
+            during the fist step.
+            The hypothetic response to the task will be subtracted, and
```

Comment:
Since score() is what users actually call instead of _score(), I added explanation of the difference to score()